interfaces/devdoc/threadpool_requirements.md

```diff
@@ -31,6 +31,12 @@ MOCKABLE_FUNCTION(, THANDLE(THREADPOOL), threadpool_create, EXECUTION_ENGINE_HAN
 MOCKABLE_FUNCTION(, int, threadpool_open, THANDLE(THREADPOOL), threadpool);
 MOCKABLE_FUNCTION(, void, threadpool_close, THANDLE(THREADPOOL), threadpool);
 
+MOCKABLE_FUNCTION(, PVOID, threadpool_create_work_item, THANDLE(THREADPOOL), threadpool, THREADPOOL_WORK_FUNCTION, work_function, PVOID, work_function_context);
+
+MOCKABLE_FUNCTION(, int, threadpool_schedule_work_item, THANDLE(THREADPOOL), threadpool, PVOID, work_item_context);
+
+MOCKABLE_FUNCTION(, void, threadpool_work_context_destroy, PVOID, work_item_context);
```

Comment:
>threadpool_work_context_destroy [](http://example.com/codeflow?start=26&length=31)

poetry: should likely be `threadpool_destroy_work_item`

Reply:
I'm not sure I follow...we need to have some function exposed here to do that cleanup, and my comment was just that the name (poetry) of this function should be `threadpool_destroy_work_item` to match the other function names.

Reply:
I am just saying that work_item aka PTP_WORK is different than WORK_ITEM_CONTEXT. Currently in the `on_work_callback`, both PTP_WORK and WORK_ITEM_CONTEXT is getting freed for every scheduling call. We will have to remove the below statements and free them. In the new implementation PTP_WORK will be saved inside WORK_ITEM_CONTEXT, hence I would call the function poetically as `threadpool_destroy_work_item_context` instead of just `threadpool_destroy_work_item`

```


/* Codes_SRS_THREADPOOL_WIN32_01_038: [ on_work_callback shall call CloseThreadpoolWork. ]*/
CloseThreadpoolWork(work);

/* Codes_SRS_THREADPOOL_WIN32_01_039: [ on_work_callback shall free the context allocated in threadpool_schedule_work. ]*/
free(context);
```

Reply:
I think the terminology here is the problem. It's a difference between the *interface* and the *implementation*.

As a "user" of the threadpool interface, I don't care about how threadpool stores its state and I don't know what a context or ptp_work is. I just know that I create some "thing" (`threadpool_create_work_item`) and when I'm done with it I call something to clean it up (`threadpool_destroy_work_item`).

In the implementation, yes, you are going to free the ptp_work and free the context. But that's an implementation detail. You are going to return the context as `THREADPOOL_WORK_ITEM_HANDLE` and take that handle for all the other calls.

Reply:
I think adding `_item` is creating all the confusion where I feel to avoid `threadpool_schedule_work` being declared twice, we are naming `threadpool_schedule_work_item` as the new API. 
So, should we have following APIs:
`threadpool_create_work`   <-- Create everything needed make scheduling work
`threadpool_schedule_work`
`threadpool_destroy_work` <--- Destroy everything that was initialized to make scheduling work
`threadpool_schedule_work_item`

For now, can we create `threadpool_schedule_work_item` as an additional new API and later when we remove `threadpool_schedule_work`, we just rename `threadpool_schedule_work_item` to `threadpool_schedule_work`.

I feel we should introduce versioning of the datastructures starting with `WORK_ITEM_CONTEXT` having a version field where version 1 is the earlier implementation and version 2 is the latest and initialized to 2 by Queue Processor

Reply:
I think either name is fine, but we should not rename `threadpool_schedule_work` right now so we don't have to update the code everywhere in all the repos.

Personally I don't see a problem with
`threadpool_create_work_item`
`threadpool_schedule_work_item`
`threadpool_destroy_work_item`

Because we are creating a "work item" which can be scheduled.

But sure, we could do a future rename to remove the "item" part.

I'm not sure we gain much with versioning the internal context structure here though. It sounds like it would add extra complexity and shouldn't be needed (it's not something we are persisting or transferring over the network where versioning would be helpful for having a mix of old/new code).